flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java

```diff
@@ -0,0 +1,218 @@
+		this.jobResultFuture = new CompletableFuture<>();
+
+		FutureUtils.assertNoException(this.jobManagerRunnerFuture.handle((jobManagerRunner, throwable) -> {
+			// JM has been initialized, or the initialization failed
+			synchronized (lock) {
+				if (throwable == null) {
+					if (cancellationFuture != null) {
+						Preconditions.checkState(jobStatus == JobStatus.CANCELLING);
+						log.warn(
+							"JobManager initialization has been cancelled for {}. Stopping JobManager.",
+							jobGraph.getJobID());
+
+						CompletableFuture<Acknowledge> cancelJobFuture = jobManagerRunner.getJobMasterGateway().thenCompose(
+							gw -> gw.cancel(Time.seconds(60))); // TODO set better timeout
+						cancelJobFuture.whenComplete((ack, cancelThrowable) -> {
+							jobStatus = JobStatus.CANCELED;
+						});
+						// forward our cancellation future
+						FutureUtils.forward(cancelJobFuture, cancellationFuture);
+						// the JobManager will complete the job result future with state CANCELLED.
+						FutureUtils.forward(jobManagerRunner.getResultFuture(), jobResultFuture);
+					} else {
+						jobStatus = JobStatus.RUNNING; // this status should never be exposed from the DispatcherJob. Only used internally for tracking runnin state
+						FutureUtils.forward(jobManagerRunner.getResultFuture(), jobResultFuture);
+					}
+				} else { // failure during initialization
+					jobStatus = JobStatus.FAILED;
+					if (submissionType == SubmissionType.RECOVERY) {
+						jobResultFuture.completeExceptionally(throwable);
+					} else {
+						jobResultFuture.complete(ArchivedExecutionGraph.createFromFailedInit(
+							jobGraph,
+							throwable,
+							jobStatus,
+							initializationTimestamp));
+					}
+				}
+			}
+			return null;
+		}));
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> getResultFuture() {
+		return jobResultFuture;
+	}
+
+	public CompletableFuture<JobDetails> requestJobDetails(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJobDetails(
+					timeout));
+			} else {
+				int[] tasksPerState = new int[ExecutionState.values().length];
+				return CompletableFuture.completedFuture(new JobDetails(
+					jobGraph.getJobID(),
+					jobGraph.getName(),
+					initializationTimestamp,
+					0,
+					0,
+					jobStatus,
+					0,
+					tasksPerState,
+					jobGraph.getVerticesAsArray().length));
+			}
+		}
+	}
+
+	public CompletableFuture<Acknowledge> cancel(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.cancel(
+					timeout));
+			} else {
+				return cancelInternal();
+			}
+		}
+	}
+
+	private CompletableFuture<Acknowledge> cancelInternal() {
+		Preconditions.checkState(Thread.holdsLock(lock));
+		jobStatus = JobStatus.CANCELLING;
+		this.cancellationFuture = new CompletableFuture<>();
```

Comment:
This looks a bit cumbersome. Why not doing something like `jobManagerRunnerFuture.thenCompose(runner -> runner.getJobMasterGateway().cancel())`?